### PR TITLE
fix for /bin/cc, month off by one, and some readability clarifications

### LINKED ISSUE #15: Kernel panic when running gzip or cc from the 3.51 Development Set
> From: Robert Masterson on Mon, May 12, 2014 at 10:18 PM
>
> Also, UNIX crashes with a panic in the emulator when I run the precompiled gzip or the cc from the 3.51 Development set.
> The input to either program does not matter--the crash always occurs as long as there is some input.
> This behavior got me thinking that there may be an error in shared memory emulation.  As you know, the 3b1 VM system has 512KB dedicated to shared memory, which some programs use, and some do not.
> Is it possible that there is a bug in the memory system relating to the shared memory segment?  (I am just guessing that cc and gzip make use of it, but *something* distinguishes these programs from those that do not cause a panic.)

### PR DESCRIPTION
**Pull request by agentbooth**
Hey Phil.  After many hours of investigation I finally got 'cc' working!  There were some issues with 32-bit writes straddling two pages which does occur. The ACCESS_CHECK_WR for 32-bits has some issues so I just replaced with two 16-bit access checks. This seems to work fine and therefore checks the first page and then the second page in a 32-bit write straddling pages. This should allow both or either page to pagefault. I think proper pagefaulting wasn't occurring in this scenario which somehow must have been triggered by cc.

### CODE REVIEW

File: src/memory.c [hunk 1]
```diff
@@ -166,22 +166,26 @@ MEM_STATUS checkMemoryAccess(uint32_t addr, bool writing, bool dma)/*{{{*/
  * 			return-with-no-value. UGH!
  */
 /*{{{ macro: ACCESS_CHECK_WR(address, bits)*/
+// TODO: 32-bit check is broken, problem with 32-bit write straddling 2 pages, proper page faults may not occur
```
Comment (philpem) [inline src/memory.c:169]: Stray TODO. Is this resolved now?

File: src/memory.c [hunk 2]
```diff
@@ -225,6 +229,8 @@ MEM_STATUS checkMemoryAccess(uint32_t addr, bool writing, bool dma)/*{{{*/
  * 			return-with-no-value. UGH!
  */
 /*{{{ macro: ACCESS_CHECK_RD(address, bits)*/
+// TODO: 32-bit check is broken, problem with 32-bit read straddling 2 pages, if needed, 2nd page PF may not occur
```
Comment (philpem) [inline src/memory.c:232]: Another TODO comment - is this resolved?

File: src/memory.c [hunk 3]
```diff
@@ -800,18 +806,21 @@ uint32_t m68k_read_memory_32(uint32_t address)/*{{{*/
 		address |= 0x800000;
 
 	// Check access permissions
-	ACCESS_CHECK_RD(address, 32);
+	ACCESS_CHECK_RD(address, 16);
+	uint32_t addr2 = address + 2;
```
Comment (philpem) [inline src/memory.c:810]: Why the additional temporary variable here?

File: src/memory.c [hunk 4]
```diff
@@ -967,14 +976,19 @@ void m68k_write_memory_32(uint32_t address, uint32_t value)/*{{{*/
 		address |= 0x800000;
 
 	// Check access permissions
-	ACCESS_CHECK_WR(address, 32);
+	ACCESS_CHECK_WR(address, 16);
+	uint32_t addr2 = address + 2;
```
Comment (philpem) [inline src/memory.c:980]: Why the additional temporary variable here?

**Comment by philpem** [comment]
Wow. I completely overlooked that the 68K can do word-aligned 32-bit loads and stores. I must have somehow got it into my head that you could only store 32bit values on a 4-byte aligned address, but the real requirement is that the address be even. Nice spot!

I've added a couple of minor review comments, and you've highlighted a couple of things I need to look into regarding the ACCESS_CHECK macros. (my preference would be to refactor them into static inlines)

For my own reference -- when merged this would close #15 .

**Review by philpem — CHANGES_REQUESTED** [review]

**Comment by agentbooth** [comment]
I think specifically what was happening was that in the ACCESS_CHECK_WR, if the 2nd word needed to page fault, which would have been detected by the address+3 check, the address to page fault wasn't updated (e.g. to address+3), and then the first page ("address") was pagefaulting.  But this was already in present, so that's what triggered the kernel panic. I looked at BSD source for the virtual memory (https://minnie.tuhs.org/cgi-bin/utree.pl?file=4.2BSD/usr/src/sys/sys/vm_page.c) to see what exactly triggered the "panic (pagein)" and it was when a page was already present -- that shouldn't be the case when trying to bring in a new page.

File: src/memory.c (hunk 4 above)
Comment (agentbooth) [reply src/memory.c:980]: I couldn't get the macro to work when I tried ACCESS_CHECK_WR((address+2), 16) so just used that temp variable.

File: src/memory.c (hunk 3 above)
Comment (agentbooth) [reply src/memory.c:810]: Same issue here, couldn't get ACCESS_CHECK_RD((address+2), 16) to work with the macro. If there's a way to get that to work without the temp variable, just let me know. I didn't spend much time messing with it.

File: src/memory.c (hunk 1 above)
Comment (agentbooth) [reply src/memory.c:169]: This is not resolved. The issue is that if the first word/page pagefaults, the pagefault is correctly triggered for the first page. However, the 2nd word (potentially on a different page) never gets checked for a fault (as far as I can tell). It works properly if the 1st page is already present, then the 2nd page can PF and be brought in correctly. A solution could be to do the _ACCESS_CHECK_WR_BYTE(address + 2) after the "if (fault)" check (and potential first page PF). But I preferred the simplicity of just calling the ACCESS_CHECK_WR twice with (address, 16) and then with (address+2, 16). What's your preference on how to handle? I like your idea of refactoring with static inlines as I found the ACCESS_CHECK _ACCESS_CHECK macros a little hard to follow.

File: src/memory.c (hunk 2 above)
Comment (agentbooth) [reply src/memory.c:232]: no, same issue as the ACCESS_CHECK_WR -- if first word/page PF's, I don't think 2nd word/page is ever checked for PF. Which could present an issue when 32-bit read straddles two pages -- and both need to PF (which may be rare).

**Comment by philpem** [comment]
It sounds like the issue is one of hardware emulation then. The 68k has a 16-bit data bus, so will do a 32bit write as two 16bit ones. I think the priority is probably to get this bug fixed, then look at refactoring later. I'll okay this for merge and open an issue for the refactor (assuming there isn't one already)
